### ISSUE TITLE
Cast Overflow (74219418)

### ISSUE THREAD

**Issue by Google-Autofuzz** [mm2/Little-CMS#155]
Hello Little-CMS team,

As part of our fuzzing efforts at Google, we have identified an issue affecting
Little-CMS (tested with revision * master da9a3290f34c83e6ba2c4ea0f29c7cdce94c2b16).

To reproduce, we are attaching a Dockerfile which compiles the project with
LLVM, taking advantage of the sanitizers that it offers. More information about
how to use the attached Dockerfile can be found here:
https://docs.docker.com/engine/reference/builder/

TL;DR instructions:
[artifacts_74219418.zip](https://github.com/mm2/Little-CMS/files/1793400/artifacts_74219418.zip)
* `mkdir project`
* `cp Dockerfile.Little-CMS /path/to/project/Dockerfile`
* `docker build --no-cache /path/to/project`
* `docker run --cap-add=SYS_PTRACE -it image_id_from_docker_build`

From another terminal, outside the container:
`docker cp /path/to/attached/reproducer running_container_hostname:/fuzzing/reproducer`
(reference: https://docs.docker.com/engine/reference/commandline/cp/)

And, back inside the container:
`/fuzzing/repro.sh /fuzzing/reproducer`

Alternatively, and depending on the bug, you could use gcc, valgrind or other
instrumentation tools to aid in the investigation. The sanitizer error that we
encountered is here:

```
INFO: Seed: 1545458220
INFO: Loaded 0 modules (0 guards): 
/fuzzing/fuzzer: Running 1 inputs 500 time(s) each.
Running: /tmp/poc-24d7a5e514c2ebec2ec918f1d17c8ac304803628f902f789033dd53f8c909eab
cmsgamma.c:1366:12: runtime error: 4.25136e+76 is outside the range of representable values of type 'float'
    #0 0x7f2e0fe690ce in cmsEvalToneCurveFloat /fuzzing/Little-CMS/src/cmsgamma.c:1366:12
    #1 0x7f2e0fe7f0c5 in EvaluateCurves /fuzzing/Little-CMS/src/cmslut.c:182:18
    #2 0x7f2e0fe82fb0 in _LUTevalFloat /fuzzing/Little-CMS/src/cmslut.c:1360:15
    #3 0x7f2e0fecb7c0 in OptimizeByComputingLinearization /fuzzing/Little-CMS/src/cmsopt.c:1112:9
    #4 0x7f2e0fec6f5d in _cmsOptimizePipeline /fuzzing/Little-CMS/src/cmsopt.c:1950:17
    #5 0x7f2e0fe9f043 in AllocEmptyTransform /fuzzing/Little-CMS/src/cmsxform.c:819:15
    #6 0x7f2e0fe9ea82 in cmsCreateExtendedTransform /fuzzing/Little-CMS/src/cmsxform.c:1075:13
    #7 0x7f2e0fe9ff6b in cmsCreateMultiprofileTransformTHR /fuzzing/Little-CMS/src/cmsxform.c:1175:12
    #8 0x7f2e0fea00c0 in cmsCreateTransformTHR /fuzzing/Little-CMS/src/cmsxform.c:1216:12
    #9 0x7f2e0fea0112 in cmsCreateTransform /fuzzing/Little-CMS/src/cmsxform.c:1226:12
    #10 0x44d619 in LLVMFuzzerTestOneInput /fuzzing/security-research-pocs/autofuzz/lcms_fuzz.cc:17:16
    #11 0x436c0e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/fuzzing/fuzzer+0x436c0e)
    #12 0x42bd5e in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/fuzzing/fuzzer+0x42bd5e)
    #13 0x430267 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/fuzzing/fuzzer+0x430267)
    #14 0x42ba7b in main (/fuzzing/fuzzer+0x42ba7b)
    #15 0x7f2e0eba32b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)
    #16 0x406609 in _start (/fuzzing/fuzzer+0x406609)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior cmsgamma.c:1366:12 in 
```

We will gladly work with you so you can successfully confirm and reproduce this
issue. Do let us know if you have any feedback surrounding the documentation.

Once you have reproduced the issue, we'd appreciate to learn your expected
timeline for an update to be released. With any fix, please attribute the report
to "Google Autofuzz project".

Don't hesitate to let us know if you have any questions!

Google AutoFuzz Team

**Comment by mm2** [comment]
Hi,I think with this report we reached a point that requires roles to be clarified. Please note I maintain this project is just for fun, and I get NO INCOME from it, so it is nice to have it clean, but this is just for my pleasure.Then, as many code doing complex math calculations, it is possible than intermediate results would overflow,  specially if the input is crafted to provoke so. Having said this, I would like to reaffirm on my purpose of take security very seriously. Any issue that would demonstrate to be a threat because allows code injection or similar danger would be fixed ASAP. And  overflows on integers / floats will not be fixed unless proven to be a threat. If this is not  acceptable to you, please avoid  to use this library. Sorry about sounding jerky, but this is necessary to maintain certain order. Thanks for your understandingSincerely,MartiOn 8 Mar 2018 14:58, Google AutoFuzz Team <notifications@github.com> wrote:Hello Little-CMS team,
As part of our fuzzing efforts at Google, we have identified an issue affecting
Little-CMS (tested with revision * master da9a329).
To reproduce, we are attaching a Dockerfile which compiles the project with
LLVM, taking advantage of the sanitizers that it offers. More information about
how to use the attached Dockerfile can be found here:
https://docs.docker.com/engine/reference/builder/
TL;DR instructions:
artifacts_74219418.zip
mkdir projectcp Dockerfile.Little-CMS /path/to/project/Dockerfiledocker build --no-cache /path/to/projectdocker run --cap-add=SYS_PTRACE -it image_id_from_docker_build
From another terminal, outside the container:
docker cp /path/to/attached/reproducer running_container_hostname:/fuzzing/reproducer
(reference: https://docs.docker.com/engine/reference/commandline/cp/)
And, back inside the container:
/fuzzing/repro.sh /fuzzing/reproducer
Alternatively, and depending on the bug, you could use gcc, valgrind or other
instrumentation tools to aid in the investigation. The sanitizer error that we
encountered is here:
INFO: Seed: 1545458220
INFO: Loaded 0 modules (0 guards): 
/fuzzing/fuzzer: Running 1 inputs 500 time(s) each.
Running: /tmp/poc-24d7a5e514c2ebec2ec918f1d17c8ac304803628f902f789033dd53f8c909eab
cmsgamma.c:1366:12: runtime error: 4.25136e+76 is outside the range of representable values of type 'float'
    #0 0x7f2e0fe690ce in cmsEvalToneCurveFloat /fuzzing/Little-CMS/src/cmsgamma.c:1366:12
    #1 0x7f2e0fe7f0c5 in EvaluateCurves /fuzzing/Little-CMS/src/cmslut.c:182:18
    #2 0x7f2e0fe82fb0 in _LUTevalFloat /fuzzing/Little-CMS/src/cmslut.c:1360:15
    #3 0x7f2e0fecb7c0 in OptimizeByComputingLinearization /fuzzing/Little-CMS/src/cmsopt.c:1112:9
    #4 0x7f2e0fec6f5d in _cmsOptimizePipeline /fuzzing/Little-CMS/src/cmsopt.c:1950:17
    #5 0x7f2e0fe9f043 in AllocEmptyTransform /fuzzing/Little-CMS/src/cmsxform.c:819:15
    #6 0x7f2e0fe9ea82 in cmsCreateExtendedTransform /fuzzing/Little-CMS/src/cmsxform.c:1075:13
    #7 0x7f2e0fe9ff6b in cmsCreateMultiprofileTransformTHR /fuzzing/Little-CMS/src/cmsxform.c:1175:12
    #8 0x7f2e0fea00c0 in cmsCreateTransformTHR /fuzzing/Little-CMS/src/cmsxform.c:1216:12
    #9 0x7f2e0fea0112 in cmsCreateTransform /fuzzing/Little-CMS/src/cmsxform.c:1226:12
    #10 0x44d619 in LLVMFuzzerTestOneInput /fuzzing/security-research-pocs/autofuzz/lcms_fuzz.cc:17:16
    #11 0x436c0e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/fuzzing/fuzzer+0x436c0e)
    #12 0x42bd5e in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/fuzzing/fuzzer+0x42bd5e)
    #13 0x430267 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/fuzzing/fuzzer+0x430267)
    #14 0x42ba7b in main (/fuzzing/fuzzer+0x42ba7b)
    #15 0x7f2e0eba32b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)
    #16 0x406609 in _start (/fuzzing/fuzzer+0x406609)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior cmsgamma.c:1366:12 in 

We will gladly work with you so you can successfully confirm and reproduce this
issue. Do let us know if you have any feedback surrounding the documentation.
Once you have reproduced the issue, we'd appreciate to learn your expected
timeline for an update to be released. With any fix, please attribute the report
to "Google Autofuzz project".
Don't hesitate to let us know if you have any questions!
Google AutoFuzz Team
